### PR TITLE
Move global 'settings' variables to C++

### PR DESCRIPTION
**Pull request by paulromano**
This PR moves the definition of most 'settings' variables to C++ as well as the associated code for reading them from XML. I've also put them in a `settings` namespace which I find really useful as an identifying marker when they are used all around the code.

### CODE REVIEW

File: src/settings.cpp [hunk 1]
```diff
@@ -16,98 +24,385 @@ namespace openmc {
...
-  if (check_for_node(*root, "output")) {
+  if (!run_CE) {
+    // Scattering Treatments
+    if (check_for_node(root, "max_order")) {
+      max_order = std::stoi(get_node_value(root, "max_order"));
+    } else {
+      // Set to default of largest int - 1, which means to use whatever is
+      // contained in library.
+      // This is largest int - 1 because for legendre scattering, a value of
+      // 1 is added to the order; adding 1 to huge(0) gets you the largest
```
Comment (smharper) [inline src/settings.cpp]: Change `huge(0)` in the comment

File: src/settings.cpp [hunk 2]
```diff
@@ -16,98 +24,385 @@ namespace openmc {
...
       }
     }
   }
 
-  // Get temperature settings
-  if (check_for_node(*root, "temperature_default")) {
-    temperature_default = std::stod(get_node_value(*root, "temperature_default"));
+  // Check run mode if it hasn't been set from the command line
+  xml_node node_mode;
+  if (run_mode == -1) {
```
Comment (smharper) [inline src/settings.cpp]: Use `C_NONE` here and for the definition of `run_mode`.

File: src/settings.cpp [hunk 3]
```diff
@@ -16,98 +24,385 @@ namespace openmc {
...
+    if (check_for_node(node_base, "inactive")) {
+      n_inactive = std::stoi(get_node_value(node_base, "inactive"));
+    }
+    if (check_for_node(node_base, "generations_per_batch")) {
+      gen_per_batch = std::stoi(get_node_value(node_base, "generations_per_batch"));
+    }
+
+    // TODO: Preallocate space for keff and entropy by generation
+
+    // TODO: Read keff_trigger information
```
Comment (smharper) [inline src/settings.cpp]: This TODO is fully resolved, right?

File: src/settings.cpp [hunk 4]
```diff
@@ -16,98 +24,385 @@ namespace openmc {
...
+    if (check_for_node(node_base, "keff_trigger")) {
+      xml_node node_keff_trigger = node_base.child("keff_trigger");
+
+      if (check_for_node(node_keff_trigger, "type")) {
+        auto temp = get_node_value(node_keff_trigger, "type", true, true);
+        if (temp == "std_dev") {
+          keff_trigger.type = STANDARD_DEVIATION;
+        } else if (temp == "variance") {
+          keff_trigger.type = VARIANCE;
+        } else if ( temp == "rel_err") {
```
Comment (smharper) [inline src/settings.cpp]: Remove that extra space in front of `temp`

File: src/settings.cpp [hunk 5]
```diff
@@ -16,98 +24,385 @@ namespace openmc {
...
+      warning("<run_mode> should be specified.");
+
+      // Make sure that either eigenvalue or fixed source was specified
+      node_mode = root.child("eigenvalue");
+      if (node_mode) {
+        if (run_mode == -1) run_mode = RUN_MODE_EIGENVALUE;
+      } else {
+        node_mode = root.child("fixed_source");
+        if (node_mode) {
+          if (run_mode == -1) run_mode = RUN_MODE_FIXEDSOURCE;
```
Comment (smharper) [inline src/settings.cpp]: The `if (run_mode == -1)` here and just above are both redundant.

File: src/settings.cpp [hunk 6]
```diff
@@ -16,98 +24,385 @@ namespace openmc {
...
+      openmc_n_threads = std::stoi(get_node_value(root, "threads"));
+      if (openmc_n_threads < 1) {
+        std::stringstream msg;
+        msg << "Invalid number of threads: " << openmc_n_threads;
+        fatal_error(msg);
+      }
+      omp_set_num_threads(openmc_n_threads);
+    }
+#else
+    if (openmc_master) warning("Ignoring number of threads.");
```
Comment (smharper) [inline src/settings.cpp]: We could make this error message more helpful by saying that OpenMC was compiled without OpenMP

File: src/settings.cpp [hunk 7]
```diff
@@ -120,6 +415,268 @@ void read_settings(pugi::xml_node* root)
...
+      energy_cutoff[3] = std::stod(get_node_value(node_cutoff, "energy_positron"));
+    }
+  }
+
+  // Particle trace
+  if (check_for_node(root, "trace")) {
+    auto temp = get_node_array<int64_t>(root, "trace");
+    trace_batch    = temp.at(0);
+    trace_gen      = temp.at(1);
+    trace_particle = temp.at(2);
```
Comment (smharper) [inline src/settings.cpp]: The error message for not specifying 3 numbers is very unhelpful.  Add something that explicitly checks the number of values and prints a helpful error message if it's wrong

File: src/settings.cpp [hunk 8]
```diff
@@ -120,6 +415,268 @@ void read_settings(pugi::xml_node* root)
...
+          + temp + ".");
+      }
+    }
+
+    // Minimum energy for resonance scattering
+    if (check_for_node(node_res_scat, "energy_min")) {
+      res_scat_energy_min = std::stod(get_node_value(node_res_scat, "energy_min"));
+    }
+    if (res_scat_energy_min < 0.0) {
+      fatal_error("Lower resonance scattering energy bound is negative");
```
Comment (smharper) [inline src/settings.cpp]: Do we actually care if the lower energy bound is negative?

File: src/settings.cpp [hunk 9]
```diff
@@ -120,6 +415,268 @@ void read_settings(pugi::xml_node* root)
...
+    }
+
+    // Maximum energy for resonance scattering
+    if (check_for_node(node_res_scat, "energy_max")) {
+      res_scat_energy_max = std::stod(get_node_value(node_res_scat, "energy_max"));
+    }
+    if (res_scat_energy_max < res_scat_energy_min) {
+      fatal_error("Upper resonance scattering energy bound is below the"
+        "lower resonance scattering energy bound.");
```
Comment (smharper) [inline src/settings.cpp]: Add a space

File: src/settings.cpp (hunk 3 above)
Comment (paulromano) [reply src/settings.cpp]: Yup, I'll remove the TODO.

File: src/settings.cpp (hunk 8 above)
Comment (paulromano) [reply src/settings.cpp]: It wouldn't really cause problems, but nevertheless it doesn't make sense so I'm in favor of just leaving the error message.

**Comment by smharper** [comment]
Looks good.  Merged!

**Pull request merged**